### ISSUE TITLE
Rework of DofDistributedPrimaryField

### ISSUE THREAD

**Issue by Micket** [oofem/oofem#33]
I think, with some rework of primaryfield, that dofdistributedprimaryfield should be the  only option. In addition, it should take over ownership of the method: "giveUnknownDictHashIndx".

However, one needs to take a close look at memory overhead here as well. The "Dictionary" datatype is bulky and introduces cachemisses. 
We should  consider having "DofDistributedPrimaryField" store the values instead (note, different from the vectors in PrimaryField, as these rely on equation numbering, and can't really store prescribed values).

Alternatively, they can stay in the MasterDof, but Dictionary should be replaced with a "small string" optimized vector, similar to  We should have a "small string" optimized vector 

In summary:
1. DofDistributedPrimaryField should own all responsibility of how the data is stored for each dof.
2. We should have more types of Fields for various types of time integration, and they should store 
3. a. MasterDof::Dictionary should be replaced with something more cache friendly. Perhaps boost::small_vector for storing 2 or 3 doubles values.
3. b. Have all this data ownership in a type of map inside DofDistributedPrimaryField (which then no longer actually would be "dof-distributed"). This may be the most cache-friendly, but, we should only consider this if it actaully matters performance-wise (I suspect it would not, at least, we have more  important performance considerations to deal with first).
